File: backend/app/schemas/games_play.py
"""Live games — student play schemas (Live Games Phase 2C, 8-8)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.services.games.avatars import AVATAR_PRESETS


def _validate_avatar(v: str | None) -> str | None:
    if v is not None and v not in AVATAR_PRESETS:
        raise ValueError(f"'{v}' is not a valid avatar — choose one of {sorted(AVATAR_PRESETS)}")
    return v


class JoinRunIn(BaseModel):
    avatar: str | None = None  # None = initials fallback (world-class rework: icon presets only, no photo)

    _validate_avatar = field_validator("avatar")(_validate_avatar)


class UpdateParticipantProfileIn(BaseModel):
    """Lobby-only nickname/avatar override for one game — distinct from the
    profile-level nickname reroll (D2, weekly-cooldown-limited): this is a
    lighter-weight, uncapped, per-game display override, per D18."""
    nickname: str = Field(min_length=1, max_length=64)
    avatar: str | None = None

    _validate_avatar = field_validator("avatar")(_validate_avatar)

    @field_validator("nickname")
    @classmethod
    def _strip_nickname(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("nickname can't be blank")
        return v


class ParticipantOut(BaseModel):
    id: UUID
    run_id: UUID
    nickname: str
    avatar: str | None
    joined_at: datetime | None


class SubmitAnswerIn(BaseModel):
    selected_option_index: int | None
    elapsed_seconds: float


class AnswerAckOut(BaseModel):
    """Private — sent only in the HTTP response and over the participant's
    own channel, never the shared run channel (D9's leaderboard privacy
    extends to the answer breakdown too, since a wrong-answer 0 next to a
    correct-answer 178 would out who answered what)."""
    is_correct: bool
    points_awarded: int
    base_points: int
    speed_bonus: int
    streak: int


class MyScoreOut(BaseModel):
    score: int
    streak: int


class StudentLeaderboardEntryOut(BaseModel):
    participant_id: UUID
    nickname: str
    avatar: str | None
    score: int
    is_me: bool


class JoinableRunOut(BaseModel):
    run_id: UUID
    assignment_id: UUID
    game_title: str
    status: str
    session_title: str | None
    session_date: str

File: frontend/src/api/games_play.ts
/** Live game student play API (Live Games Phase 2C, 8-8) — thin wrapper
 * over `/games/play/*`. Types mirror `schemas/games_play.py`. */
import { api } from "@/api/client";
import type { GameRun, PublicQuestion, RosterEntry } from "@/api/games_live";

export interface JoinableRun {
  run_id: string;
  assignment_id: string;
  game_title: string;
  status: "lobby" | "live";
  session_title: string | null;
  session_date: string;
}

export interface Participant {
  id: string;
  run_id: string;
  nickname: string;
  avatar: string | null;
  joined_at: string | null;
}

export interface AnswerAck {
  is_correct: boolean;
  points_awarded: number;
  base_points: number;
  speed_bonus: number;
  streak: number;
}

export interface MyScore {
  score: number;
  streak: number;
}

export interface StudentLeaderboardEntry {
  participant_id: string;
  nickname: string;
  avatar: string | null;
  score: number;
  is_me: boolean;
}

export const getJoinableRunsApi = () => api.get<JoinableRun[]>("/games/play/joinable").then((r) => r.data);

export const joinRunApi = (runId: string, avatar?: string | null) =>
  api.post<Participant>(`/games/play/runs/${runId}/join`, { avatar: avatar ?? null }).then((r) => r.data);

export const updateMyProfileApi = (runId: string, nickname: string, avatar: string | null) =>
  api.patch<Participant>(`/games/play/runs/${runId}/me`, { nickname, avatar }).then((r) => r.data);

export const getPlayRosterApi = (runId: string) =>
  api.get<RosterEntry[]>(`/games/play/runs/${runId}/roster`).then((r) => r.data);

export const getPlayRunApi = (runId: string) => api.get<GameRun>(`/games/play/runs/${runId}`).then((r) => r.data);

export const getPlayQuestionApi = (runId: string) =>
  api.get<PublicQuestion>(`/games/play/runs/${runId}/question`).then((r) => r.data);

export const submitAnswerApi = (runId: string, selectedOptionIndex: number | null, elapsedSeconds: number) =>
  api.post<AnswerAck>(`/games/play/runs/${runId}/answer`, {
    selected_option_index: selectedOptionIndex, elapsed_seconds: elapsedSeconds,
  }).then((r) => r.data);

export const getMyScoreApi = (runId: string) => api.get<MyScore>(`/games/play/runs/${runId}/my-score`).then((r) => r.data);

export const getStudentLeaderboardApi = (runId: string) =>
  api.get<StudentLeaderboardEntry[]>(`/games/play/runs/${runId}/leaderboard`).then((r) => r.data);
